beacon-chain/blockchain/forkchoice/service.go

```diff
@@ -0,0 +1,460 @@
+	// Make block root as bad to reject in sync.
+	postState, err := state.ExecuteStateTransition(s.ctx, preState, b)
+	if err != nil {
+		if err := s.db.DeleteBlock(s.ctx, root); err != nil {
+			return errors.Wrap(err, "could not delete bad block from db")
+		}
+		return errors.Wrap(err, "could not execute state transition")
+	}
+
+	if err := s.db.SaveState(s.ctx, postState, root); err != nil {
+		return errors.Wrap(err, "could not save state")
+	}
+
+	// Update justified check point.
+	if postState.CurrentJustifiedCheckpoint.Epoch > s.justifiedCheckpt.Epoch {
+		s.justifiedCheckpt = postState.CurrentJustifiedCheckpoint
+	}
+
+	// Update finalized check point.
+	// Prune the block cache and helper caches on every new finalized epoch.
+	if postState.FinalizedCheckpoint.Epoch > s.finalizedCheckpt.Epoch {
+		helpers.ClearAllCaches()
+		s.finalizedCheckpt.Epoch = postState.FinalizedCheckpoint.Epoch
+	}
+
+	// Log epoch summary before the next epoch.
+	if helpers.IsEpochStart(postState.Slot) {
+		logEpochData(postState)
+	}
+
+	return nil
+}
+
+// OnAttestation is called whenever an attestation is received, it updates validators latest vote.
+//
+// Spec pseudocode definition:
+//   def on_attestation(store: Store, attestation: Attestation) -> None:
+//    target = attestation.data.target
+//
+//    # Cannot calculate the current shuffling if have not seen the target
+//    assert target.root in store.blocks
+//
+//    # Attestations cannot be from future epochs. If they are, delay consideration until the epoch arrives
+//    base_state = store.block_states[target.root].copy()
+//    assert store.time >= base_state.genesis_time + compute_start_slot_of_epoch(target.epoch) * SECONDS_PER_SLOT
+//
+//    # Store target checkpoint state if not yet seen
+//    if target not in store.checkpoint_states:
+//        process_slots(base_state, compute_start_slot_of_epoch(target.epoch))
+//        store.checkpoint_states[target] = base_state
+//    target_state = store.checkpoint_states[target]
+//
+//    # Attestations can only affect the fork choice of subsequent slots.
+//    # Delay consideration in the fork choice until their slot is in the past.
+//    attestation_slot = get_attestation_data_slot(target_state, attestation.data)
+//    assert store.time >= (attestation_slot + 1) * SECONDS_PER_SLOT
+//
+//    # Get state at the `target` to validate attestation and calculate the committees
+//    indexed_attestation = get_indexed_attestation(target_state, attestation)
+//    assert is_valid_indexed_attestation(target_state, indexed_attestation)
+//
+//    # Update latest messages
+//    for i in indexed_attestation.custody_bit_0_indices + indexed_attestation.custody_bit_1_indices:
+//        if i not in store.latest_messages or target.epoch > store.latest_messages[i].epoch:
+//            store.latest_messages[i] = LatestMessage(epoch=target.epoch, root=attestation.data.beacon_block_root)
+func (s *Store) OnAttestation(a *ethpb.Attestation) error {
```

Comment:
Same here @terencechain, this func does too much. It will also simplify tests and make for better maintainability if we abstract into smaller pieces.

Reply:
yep